### ISSUE TITLE
[Bug]: Storybook freezes when trying to change args through Controls

### ISSUE THREAD

**Issue by SwimmingPolar** [storybookjs/storybook#19695]
### Describe the bug

Not really sure if it's a bug or me just doing things in a wrong way but I have some leads why storybook freezes.
I think it happens when trying to provide react-router-dom's memory router to story through component decorators.

Here's a repo to the code.
https://github.com/SwimmingPolar/storybook-freeze

``` typescript
export default {
  title: "Components/Navbar",
  component: Navbar,
  parameters: {
    layout: "fullscreen",
  },
  args: { menu },
  argTypes: {
    menu: {
      type: {
        name: "array",
      } as SBArrayType,
    },
    hide: {
      control: {
        type: "boolean",
      },
    },
  },
  decorators: [(Story) => <RouterProvider router={memoryRouter(<Story />)} />],
} as ComponentMeta<typeof Navbar>;

const Template: ComponentStory<typeof Navbar> = (args) => <Navbar {...args} />;

export const Default = Template.bind({});
Default.storyName = "Show Menu";

export const Hide = Template.bind({});
Hide.storyName = "Hide Menu";
Hide.args = { hide: true };
```

It's just simple navbar rendering a list of menu items it took as props and decides to show the menu or not base on `hide` prop.
When I click on the menu, the route changes and the content as well. But say if I want to change `hide` to true or false through Controles panel below, the storybook ui or the entire browser tab stops as if it's awaiting a promise or the thread is taken by synchronous job.

I'm not really sure if it's right way to provide providers in storybook but this is how I did it.

_component decorators_
``` typescript
decorators: [(Story) => <RouterProvider router={memoryRouter(<Story />)} />],
```
_Router Factory function_
``` typescript
const RouteWrapper = styled.div``;

const Route = styled.div`
  height: 1500px;
`;

export const memoryRouter = (children: ReactElement) =>
  createMemoryRouter(
    [
      {
        path: "/",
        element: <Layout>{children}</Layout>,
        children: [
          {
            element: (
              // To add extra styles to 'content' area of the entire layout
              <RouteWrapper>
                <Outlet />
              </RouteWrapper>
            ),
            children: [
              {
                index: true,
                path: "/home",
                element: <Route>Home Content</Route>,
              },
              {
                path: "/about",
                element: <Route>About Content</Route>,
              },
              {
                path: "*",
                element: <Route>404 Not Found</Route>,
              },
            ],
          },
        ],
      },
    ],
    { initialEntries: ["/home"], initialIndex: 0 }
  );

```

### To Reproduce

```shell
https://github.com/SwimmingPolar/storybook-freeze
```

**Comment by SwimmingPolar** [comment]
![image](https://user-images.githubusercontent.com/53928959/199144932-c2218778-34cf-4211-8dbd-ccb29823a7ba.png)

It looks like components are not unmounting properly when attempt to change values from Controls panel. It jumps straight into fresh rendering of the entire component tree. Maybe that's what's causing the freeze?

**Comment by codebutler** [comment]
This might be a duplicate of #17720. 

See my comment over there, does that fix your problem? 
https://github.com/storybookjs/storybook/issues/17720#issuecomment-1466725964

**Comment by shilman** [comment]
Closing as duplicate to https://github.com/storybookjs/storybook/issues/17720. Please feel free to reopen if that's not the case.

**Comment by SwimmingPolar** [comment]
> 

Yes it does fix the issue. Thanks for the heads up! It must be some kind of bug. Hope storybook team fixes this by default.

**Comment by SwimmingPolar** [comment]
> This might be a duplicate of #17720.
> 
> See my comment over there, does that fix your problem? [#17720 (comment)](https://github.com/storybookjs/storybook/issues/17720#issuecomment-1466725964)

Yes it does fix the issue. Thanks for the heads up! It must be some kind of bug. Hope storybook team fixes this by default.
